### PR TITLE
[ELY-1014] DigestAuthenticationMechanism fixes for incomplete config

### PR DESCRIPTION
**Pull request by honza889**
* fix for situation when default realm is configured, but available realms are not
* fix for wrong realm name in users clear property file / realms with two-way passwords without realm name consistent with auth mechanism (digest from two-way first)

Now for working digest auth it is sufficient to have corresponding realm names in one of following:
* a) web.xml AND user property file
* b) web.xml AND mechanism in http-authentication-factory

ADDED security fix:
* Removed defaulting realm to HTTP Host - was allowing to spoof unallowed realm from client

### CODE REVIEW

File: src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java [hunk 1]
```diff
@@ -818,6 +818,7 @@ private void handleOne(final Callback[] callbacks, final int idx) throws IOExcep
                 } else if (callback instanceof CredentialCallback) {
                     final CredentialCallback credentialCallback = (CredentialCallback) callback;
 
+                    // TODO: should consider realm from RealmCallback
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: I assume you're referring to org.wildfly.security.password.spec.DigestPasswordAlgorithmSpec here?

Comment (honza889) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: yes, it sounds like a way by which it could work

File: src/main/java/org/wildfly/security/http/impl/DigestAuthenticationMechanism.java [hunk 2]
```diff
@@ -113,7 +110,7 @@ public String getMechanismName() {
 
     @Override
     public void evaluateRequest(final HttpServerRequest request) throws HttpAuthenticationException {
-        final String realmName = selectRealm(request);
+        final String defaultRealm = selectDefaultRealm();
```
Comment (darranl) [inline src/main/java/org/wildfly/security/http/impl/DigestAuthenticationMechanism.java]: We may want to double check the available realms as the mechanism is actually created - at that point we should be able to identify if we have realms available for the mechanism to function or fail to be created if no realms were available.

The previous code only waited until evaluateRequest so we could fallback to using the Host header.

**Comment by honza889** [comment]
(should be run wildfly+wildfly-core tests)

File: src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java (hunk 1 above)
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: If we wanted to do this, I think the right place to do this would be in org.wildfly.security.sasl.digest.AbstractDigestMechanism#getPredigestedSaltedPassword.  The problem is that the parameters depend on the selected realm, and the realm callback is handled at the same time as the credential callback.  So, it might not actually be possible.

Comment (honza889) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: Looking for way how to obtain PasswordCredential with specified realm. But to obtain TwoWayPassword and convert to digest password of given realm would be maybe also appropriate...
In both cases I think it should be in standalone PR, so only TODO keeping here.

Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: +1

**Comment by honza889** [comment]
hold, brokes wildfly-core/rbac tests

**Comment by honza889** [comment]
Replaced by https://github.com/wildfly-security/wildfly-elytron/pull/746
